### PR TITLE
feat: Add push_validity_into_children methods to StructArray

### PR DESCRIPTION
**Pull request by amorynan**
For solving : https://github.com/vortex-data/vortex/issues/3859
Add methods to push struct-level validity into child fields:
- push_validity_into_children(preserve_struct_validity: bool)
- push_validity_into_children_default() - convenience method with preserve=false

The functionality propagates null information from struct level down to individual fields, with options to preserve or remove the struct-level validity.

Includes comprehensive tests covering all scenarios:
- preserve_struct_validity = true
- preserve_struct_validity = false (default)
- no nulls edge case

### CODE REVIEW

File: vortex-array/src/arrays/struct_/array.rs [hunk 1]
```diff
@@ -451,4 +452,112 @@ impl StructArray {
...
+    /// // pushed_no_struct.fields()[0] now has nulls at position 1
+    /// // pushed_no_struct.fields()[1] now has nulls at position 1
+    /// // pushed_no_struct.validity is AllValid
+    /// ```
+    /// Push validity into children with default behavior (preserve_struct_validity = false).
+    ///
+    /// This is equivalent to calling `push_validity_into_children(false)`.
+    pub fn push_validity_into_children_default(&self) -> VortexResult<Self> {
+        self.push_validity_into_children(false)
+    }
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs]: I don't think we need to have this method, the one below should be sufficient since a the `bool` impl of `Default` is already `false`.

File: vortex-array/src/arrays/struct_/array.rs [hunk 2]
```diff
@@ -451,4 +452,112 @@ impl StructArray {
...
+
+
+    pub fn push_validity_into_children(&self, preserve_struct_validity: bool) -> VortexResult<Self> {
+        use crate::compute::mask;
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs]: Can you move this import to the top?

File: vortex-array/src/arrays/struct_/array.rs [hunk 3]
```diff
@@ -451,4 +452,112 @@ impl StructArray {
...
+
+        // Get the struct-level validity mask
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs]: Nit: periods at the ends of sentences.

**Review by connortsui20 — CHANGES_REQUESTED** [review]
I think this is a good first step, though I'm trying to figure out how we want to implement this behavior for the new operator world we are migrating to.

@gatesn mentioned this in the original issue:

> Since this issue was filed, we now have a "Mask" expression that essentially performs intersection with validity.
>
> So you could take a look at how you might take the validity from the struct array itself, and wrap up each child field in a mask expression (see builtins.rs), before constructing a new StructArray with or without the previous validity. These two are different behaviors.

This PR implements this feature correctly for the old (current) world, but we will likely want to implement the behavior described above very very soon, so soon that it might not even be worth merging this right now? @gatesn do you have any thoughts?

File: vortex-array/src/arrays/struct_/array.rs (hunk 3 above)
Comment (amorynan) [reply vortex-array/src/arrays/struct_/array.rs]: OK, I had fix it . Please review again.

File: vortex-array/src/arrays/struct_/array.rs (hunk 1 above)
Comment (amorynan) [reply vortex-array/src/arrays/struct_/array.rs]: Yes , I remove the default Method . Please review it .

File: vortex-array/src/arrays/struct_/array.rs (hunk 2 above)
Comment (connortsui20) [reply vortex-array/src/arrays/struct_/array.rs]: I also prefer to see `compute::mask` as the function call instead of just `mask`, but that is just my personal preference

File: vortex-array/src/arrays/struct_/array.rs [hunk 4]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+    /// // pushed.fields()[0] now has nulls at position 1
+    /// // pushed.fields()[1] now has nulls at position 1
+    /// // pushed.validity still shows row 1 as null
+    ///
+    /// // Push validity into children, removing struct validity
+    /// let pushed_no_struct = struct_array.push_validity_into_children(false).unwrap();
+    /// // pushed_no_struct.fields()[0] now has nulls at position 1
+    /// // pushed_no_struct.fields()[1] now has nulls at position 1
+    /// // pushed_no_struct.validity is AllValid
+    /// ```
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs]: I think it would be more helpful to comment what the struct array looks like instead of requiring the reader to parse what the created struct array becomes after `try_new`. And you might as well have assertions stating that specific positions are null rather than just comments.

File: vortex-array/src/arrays/struct_/array.rs [hunk 5]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+    /// use vortex_array::validity::Validity;
+    /// use vortex_array::IntoArray;
+    /// use vortex_buffer::buffer;
+    ///
+    /// // Create struct with top-level nulls
+    /// let struct_array = StructArray::try_new(
+    ///     ["a", "b"].into(),
+    ///     vec![
+    ///         buffer![1i32, 2i32, 3i32].into_array(),
+    ///         buffer![10i32, 20i32, 30i32].into_array(),
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs]: It may not be immediately obvious to someone reading this that the child buffers are non-nullable

File: vortex-array/src/arrays/struct_/array.rs [hunk 6]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+        if struct_validity_mask.all_true() {
+            return if preserve_struct_validity {
+                Ok(self.clone())
+            } else {
+                // Remove struct validity if requested.
+                Self::try_new(
+                    self.names().clone(),
+                    self.fields().clone(),
+                    self.len(),
+                    Validity::AllValid,
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs:523]: Hmmm I wonder if we should make the validity of the top-level struct array `Validity::NonNullable` instead of `AllValid` if the user requests to not preserve the struct validity. Do you have any thoughts?

File: vortex-array/src/arrays/struct_/array.rs [hunk 7]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+        });
+
+        let masked_fields: Vec<ArrayRef> = self
+            .fields()
+            .iter()
+            .map(|field| {
+                // Use the mask function to apply null positions to each field.
+                mask(field.as_ref(), &null_mask)
+            })
+            .collect::<VortexResult<Vec<_>>>()?;
```
Comment (connortsui20) [inline vortex-array/src/arrays/struct_/array.rs:542]: This inverted logic makes me think more that we should just move this to the new world now instead of merging this and then getting rid of it immediately. @gatesn Any thoughts about this?

File: vortex-array/src/arrays/struct_/array.rs (hunk 2 above)
Comment (amorynan) [reply vortex-array/src/arrays/struct_/array.rs]: done!

File: vortex-array/src/arrays/struct_/array.rs (hunk 4 above)
Comment (amorynan) [reply vortex-array/src/arrays/struct_/array.rs]: done!

File: vortex-array/src/arrays/struct_/array.rs [hunk 8]
```diff
@@ -451,4 +452,108 @@ impl StructArray {
...
+                    self.len(),
+                    Validity::AllValid,
+                )
+            };
+        }
+
+        // Apply the struct validity mask to each child field.
+        // We want to set nulls where the struct is null (i.e., where struct_validity_mask is false).
+        // So we need to invert the mask: where struct is invalid, set child to invalid.
+        let null_mask = struct_validity_mask.iter_bools(|iter| {
```
Comment (joseph-isaacs) [inline vortex-array/src/arrays/struct_/array.rs:531]: there is a  not method

File: vortex-array/src/arrays/struct_/array.rs [hunk 9]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+    ///
+    /// This method takes the top-level validity of the struct array and applies it to each child field
+    /// using a mask operation. The resulting struct array will have the struct-level nulls propagated
+    /// down to the individual fields.
+    ///
+    /// # Parameters
+    ///
+    /// * `preserve_struct_validity` - If true, the new struct array retains the original struct-level
+    ///   validity. If false, the new struct array has `Validity::AllValid` since all null information
+    ///   is now contained within the individual fields.
```
Comment (joseph-isaacs) [inline vortex-array/src/arrays/struct_/array.rs:466]: can we invert this and have a remove_struct_validity since the default (false) should be to keep it. Since that doesn't change the values in the struct whereas removing this does

File: vortex-array/src/arrays/struct_/array.rs (hunk 4 above)
Comment (joseph-isaacs) [reply vortex-array/src/arrays/struct_/array.rs]: create a two struct are compare using assert_array_eq!

File: vortex-array/src/arrays/struct_/array.rs [hunk 10]
```diff
@@ -451,4 +452,106 @@ impl StructArray {
...
+    /// // pushed.fields()[1] now has nulls at position 1
+    /// // pushed.validity still shows row 1 as null
+    ///
+    /// // Push validity into children, removing struct validity
+    /// let pushed_no_struct = struct_array.push_validity_into_children(false).unwrap();
+    /// // pushed_no_struct.fields()[0] now has nulls at position 1
+    /// // pushed_no_struct.fields()[1] now has nulls at position 1
+    /// // pushed_no_struct.validity is AllValid
+    /// ```
+    /// Push validity into children with default behavior (preserve_struct_validity = false).
```
Comment (joseph-isaacs) [inline vortex-array/src/arrays/struct_/array.rs:503]: invert please

File: vortex-array/src/arrays/struct_/array.rs (hunk 7 above)
Comment (joseph-isaacs) [reply vortex-array/src/arrays/struct_/array.rs:542]: I think we should keep this for now

File: vortex-array/src/arrays/struct_/tests.rs
Comment (joseph-isaacs) [inline vortex-array/src/arrays/struct_/tests.rs:1]: can you use `assert_arrays_eq!` to check equality

**Review by joseph-isaacs — COMMENTED** [review]
I wonder if we want to split this into two functions. 

One which pushes down the validity into all children and makes the struct non-nullable (`push_validity_into_children()`) and another that only push down validity (possibly duplicating it) (`compact_validity()`)
